src/httpserver.cpp

```diff
@@ -457,21 +475,21 @@ void StopHTTPServer()
         delete workQueue;
         workQueue = nullptr;
     }
+    {
+        LogPrint(BCLog::HTTP, "Disabling reading on current connections\n");
+        LOCK(g_http_connections_cs);
+        for (evhttp_connection* conn : g_http_connections) {
+            bufferevent* bev = evhttp_connection_get_bufferevent(conn);
+            if (bev) bufferevent_disable(bev, EV_READ);
+        }
+        g_http_connections.clear();
+    }
     if (eventBase) {
         LogPrint(BCLog::HTTP, "Waiting for HTTP event thread to exit\n");
-        // Exit the event loop as soon as there are no active events.
-        event_base_loopexit(eventBase, nullptr);
-        // Give event loop a few seconds to exit (to send back last RPC responses), then break it
-        // Before this was solved with event_base_loopexit, but that didn't work as expected in
-        // at least libevent 2.0.21 and always introduced a delay. In libevent
-        // master that appears to be solved, so in the future that solution
-        // could be used again (if desirable).
-        // (see discussion in https://github.com/bitcoin/bitcoin/pull/6990)
-        if (threadResult.valid() && threadResult.wait_for(std::chrono::milliseconds(2000)) == std::future_status::timeout) {
-            LogPrintf("HTTP event loop did not exit within allotted time, sending loopbreak\n");
-            event_base_loopbreak(eventBase);
-        }
+        event_base_loopbreak(eventBase);
         threadHTTP.join();
+        // Process remaining events.
+        event_base_dispatch(eventBase);
```

Comment:
This is necessary in order to have a clean event loop quit, which means all active events processed and there's no pending events. In some cases (long response/slow client) the timeout can cause errors on the client side because the client receives an unexpected connection close.

Ideally it would be enough:
```cpp
// Wait the event loop to quit
threadHTTP.join();
```
but in windows only this works:
```cpp
// Wait the event loop to quit in *this* thread
event_base_loopbreak(eventBase);
threadHTTP.join();
// Process remaining events.
event_base_dispatch(eventBase);
```